services/src/handlers/model_training.rs

```diff
@@ -0,0 +1,199 @@
+}
+
+/// This method forwards the collected tile data to the appropriate implementation of
+/// the actual aggregation algorithm.
+fn accumulate_tile_data(
+    tile_data: Vec<Option<f32>>,
+    accumulator: &mut Box<dyn Aggregatable<Data = Vec<f32>> + Send>,
+) {
+    let v = tile_data.into_iter().flatten().collect::<Vec<f32>>();
+
+    accumulator.aggregate(v);
+}
+
+//TODO: add a way to abort the query execution when the tasks is aborted
+/// Build ML Features from the raw data and assign feature names.
+pub async fn accumulate_raster_data(
+    feature_names: Vec<Option<String>>,
+    processors: Vec<TypedRasterQueryProcessor>,
+    query: VectorQueryRectangle,
+    ctx: &dyn QueryContext,
+    accum_variant: MachineLearningAggregator,
+) -> Result<Vec<MachineLearningFeature>> {
+    type MlData = Box<dyn Aggregatable<Data = Vec<f32>> + Send>;
+
+    let mut queries = Vec::with_capacity(processors.len());
+    let q = query.into();
+    for (i, raster_processor) in processors.iter().enumerate() {
+        queries.push(
+            call_on_generic_raster_processor!(raster_processor, processor => {
+                processor.query(q, ctx).await?
+                         .and_then(
+                            move |tile| spawn_blocking_with_thread_pool(
+                                ctx.thread_pool().clone(),
+                                move || (i, tile.convert_data_type_parallel()) ).map_err(Into::into)
+                        ).boxed()
+            }),
+        );
+    }
+
+    let mut aggregator_vec = Vec::new();
+    let n_rasters = processors.len();
+
+    // populate the aggregator vector with an aggregator for each raster
+    match accum_variant {
+        MachineLearningAggregator::Simple => {
+            (0..n_rasters).into_iter().for_each(|_| {
+                let aggregator = SimpleAggregator::<f32> { data: Vec::new() };
+                aggregator_vec.push(Box::new(aggregator) as MlData);
+            });
+        }
+        MachineLearningAggregator::ReservoirSampling => {
+            (0..n_rasters).into_iter().for_each(|_| {
+                let aggregator = ReservoirSamplingAggregator::<f32> { data: Vec::new() };
+                aggregator_vec.push(Box::new(aggregator) as MlData);
+            });
+        }
+    }
+
+    let result = select_all(queries)
+        .fold(
+            Ok(aggregator_vec),
+            |aggregator_vec: Result<Vec<MlData>>, enumerated_raster_tile| async move {
+                let mut aggregator_vec = aggregator_vec?;
+                let (i, raster_tile) = enumerated_raster_tile?;
+
+                let res = match raster_tile.grid_array {
+                    GridOrEmpty::Grid(g) => {
+                        let agg = aggregator_vec
+                            .get_mut(i)
+                            .ok_or(XgModuleError::CouldNotGetMlAggregatorRef { index: i })?;
+
+                        let tile_data: Vec<_> = g.masked_element_deref_iterator().collect();
```

Comment:
Can you maybe change the associated type `Data` in 
``` 
pub trait Aggregatable {
    type Data;
```
s.t. `Data` is the actual pixel data type instead of a Vec of it?

then

```
pub async fn accumulate_raster_data<A>(
....
where
    A: Aggregatable<Data = Vec<f32>>,
```

could be 

```
pub async fn accumulate_raster_data<A>(
....
where
    A: Aggregatable<Data = f32>,
```

and in Aggregatable

```
 fn finish(&self) -> &[Self::Data];
```

Then it should be generic again.

Reply:
c4ee563f

commit msg:
 refactor: change Aggregatable::finish to return owned data and change the associated type